pkg/blockdevice/v1alpha2/blockdevice.go

```diff
@@ -0,0 +1,178 @@
+	Object *ndmapisv1alpha1.BlockDevice
+}
+
+// BlockDeviceList holds the list of BlockDevice api
+type BlockDeviceList struct {
+	// list of blockdevices
+	ObjectList *ndmapisv1alpha1.BlockDeviceList
+}
+
+// Predicate defines an abstraction to determine conditional checks against the
+// provided block device instance
+type Predicate func(*BlockDevice) bool
+
+// predicateList holds the list of Predicates
+type predicateList []Predicate
+
+// all returns true if all the predicates succeed against the provided block
+// device instance.
+func (l predicateList) all(c *BlockDevice) bool {
+	for _, pred := range l {
+		if !pred(c) {
+			return false
+		}
+	}
+	return true
+}
+
+// HasAnnotation returns true if provided annotation key and value are present
+// in the provided block deive instance.
+func HasAnnotation(key, value string) Predicate {
+	return func(c *BlockDevice) bool {
+		val, ok := c.Object.GetAnnotations()[key]
+		if ok {
+			return val == value
+		}
+		return false
+	}
+}
+
+// TODO: fix below snippet when #1227 merged
+//// IsSparse returns true if the block device is of sparse type.
+//func IsSparse() Predicate {
+//	return func(d *BlockDevice) bool {
+//		return d.Object.GetLabels()[string(apisv1alpha1.NdmBlockDeviceTypeCPK)] == string(apisv1alpha1.TypeBlockDeviceCPV)
+//	}
+//}
+
+// IsType returns true if the block device is of type same as passed argument
+//func IsType(bdType string) Predicate {
+//	return func(d *BlockDevice) bool {
+//		return d.Object.GetLabels()[string(apisv1alpha1.NdmBlockDeviceTypeCPK)] == bdType
+//	}
+//}
+
+// IsActive returns true if the block device is active.
+func IsActive() Predicate {
+	return func(bd *BlockDevice) bool {
+		return bd.Object.Status.State == string(BlockDeviceStateActive)
+	}
+}
+
+// IsUsable returns true if this block device can be used for pool provisioning.
+// The argument usedBlockDevice is a map containing key as block device cr name and value as integer.
+// If the value of map is greater than 0 , then this corresponding block device is not usable.
+func IsUsable(usedBlockDevices map[string]int) Predicate {
+	return func(bd *BlockDevice) bool {
+		return usedBlockDevices[bd.Object.Name] == 0
+	}
+}
+
+// IsUsableNode returns true if block device of this node can be used for pool provisioning.
+// The argument usedNodes is a map containing key as node name and value as bool.
+// If the value of map is greater than false, then this corresponding node is not usable.
+func IsUsableNode(usedNodes map[string]bool) Predicate {
+	return func(bd *BlockDevice) bool {
+		return !usedNodes[bd.GetNodeName()]
+	}
+}
+
+// IsBelongToNode returns true if the block device belongs to the provided node.
+func IsBelongToNode(nodeName string) Predicate {
+	return func(bd *BlockDevice) bool {
+		return bd.GetNodeName() == nodeName
+	}
+}
+
+// IsValidPoolTopology returns true if the topology is valid.
+func IsValidPoolTopology(poolType string, diskCount int) bool {
```

Comment:
Please explain in comments what is meant by topology

Reply:
Done